return type inference broken for generic closures

### Example

In code below the `(r) => (r - v).abs` returns `int`, but the function `findi_min` expect `float`, Nim wrongly compiles that code and later C-compiler throws an error.

```nim
import sugar, sequtils

func findi_min*[T](list: openarray[T], op: (T) -> float): int =
  0

proc find_something(v: int, list: seq[int]): int =
  list.findi_min((r) => (r - v).abs)

echo find_something(1, @[1])
```

### Current Output

```
Hint: used config file '/playground/nim/config/nim.cfg' [Conf]
Hint: used config file '/playground/nim/config/config.nims' [Conf]
.........
/usercode/in.nim(1, 15) Warning: imported and not used: 'sequtils' [UnusedImport]
Hint: gcc -c  -w -fmax-errors=3   -I/playground/nim/lib -I/usercode -o /usercode/nimcache/stdlib_io.nim.c.o /usercode/nimcache/stdlib_io.nim.c [Exec]
Hint: gcc -c  -w -fmax-errors=3   -I/playground/nim/lib -I/usercode -o /usercode/nimcache/stdlib_system.nim.c.o /usercode/nimcache/stdlib_system.nim.c [Exec]
Hint: gcc -c  -w -fmax-errors=3   -I/playground/nim/lib -I/usercode -o /usercode/nimcache/@min.nim.c.o /usercode/nimcache/@min.nim.c [Exec]
/usercode/nimcache/@min.nim.c: In function 'find_something__Sy1A5qO72g9aM0AG1s9c03Iw':
/usercode/nimcache/@min.nim.c:216:86: error: incompatible type for argument 3 of 'findi_min__21HJ9aws2bHQ5Ulk4MVcpCQ'
  result = findi_min__21HJ9aws2bHQ5Ulk4MVcpCQ(list->data, (list ? list->Sup.len : 0), T1_);
                                                                                      ^~~
In file included from /usercode/nimcache/@min.nim.c:11:0:
/usercode/nimcache/@min.nim.c:178:29: note: expected 'tyProc__1un7Zt9bB5WVQaIdXFZeiNQ {aka struct <anonymous>}' but argument is of type 'tyProc__vwnqUR9cTZ8xwFKGcxd2hkw {aka struct <anonymous>}'
 N_LIB_PRIVATE N_NIMCALL(NI, findi_min__21HJ9aws2bHQ5Ulk4MVcpCQ)(NI* list, NI listLen_0, tyProc__1un7Zt9bB5WVQaIdXFZeiNQ op) {
                             ^
/playground/nim/lib/nimbase.h:252:44: note: in definition of macro 'N_NIMCALL'
 #  define N_NIMCALL(rettype, name) rettype name /* no modifier */
                                            ^~~~
Error: execution of an external program failed: 'gcc -c  -w -fmax-errors=3   -I/playground/nim/lib -I/usercode -o /usercode/nimcache/@min.nim.c.o /usercode/nimcache/@min.nim.c'
```

### Expected Output

Nim compilation error that `(r) => (r - v).abs` should have `float` return type instead of `int`.

P.S.

A similar error is that the code below compiles and runs, but it shouldn't. It should result in Nim compile-time error saying that `(r) => (r - 1)` returns `int` but the `float` expected.

```Nim
import sugar

func findi_min*[T](list: openarray[T], op: (T) -> float): int =
  0

echo @[1].findi_min((r) => (r - 1))
```

reduced without stdlib dependencies:

```nim
when true:
  func fn[T](a: T, op: proc(a: T): float) = discard
  proc main() =
    let v = 1
    proc bar(r: auto): auto = v
    fn(1, bar)
  main()
```

```
@mt12614.nim.c:168:30: error: passing 'tyProc__vwnqUR9cTZ8xwFKGcxd2hkw' to parameter of incompatible type 'tyProc__1un7Zt9bB5WVQaIdXFZeiNQ'
        fn_t4950544952_18(((NI) 1), T1_);
                                    ^~~
/Users/timothee/git_clone/nim/timn/build/nimcache/@mt12614.nim.c:132:88: note: passing argument to parameter 'op' here
N_LIB_PRIVATE N_NIMCALL(void, fn_t4950544952_18)(NI a, tyProc__1un7Zt9bB5WVQaIdXFZeiNQ op) {

```

--- a/tests/errmsgs/t16654.nim
+++ b/tests/errmsgs/t16654.nim
@@ -1,0 +1,12 @@
+discard """
+  cmd: "nim check $options $file"
+  errormsg: "type mismatch: got <int> but expected 'float'"
+"""
+
+when true: # bug #16654
+  func fn[T](a: T, op: proc(a: T): float) = discard
+  proc main() =
+    let v = 1
+    proc bar(r: auto): auto = v
+    fn(1, bar)
+  main()

